### PR TITLE
Vector count optimization

### PR DESCRIPTION
**Pull request by ivanazeljkovic**
Current implementation of method for extracting vector count is based on **querying with dummy embeddings** (empty embeddings). Since this approach is time-consuming (due to extra step - calculating similarity between dummy query vector and all vectors that satisfy metadata filter), it is necessary to make an optimization - extract vector count from the **index stats**, if possible.
In this way, all starters except _gcp-starter_ will use index stats for purpose of getting vector count, except _gcp-starter_ that will use existing approach (due to limitations on metadata filtering).

### CODE REVIEW

File: haystack/document_stores/pinecone.py [hunk 1]
```diff
@@ -188,17 +198,8 @@ def __init__(
...
-        Add all document IDs to the set of all IDs.
-        """
-        if index not in self.all_ids:
-            self.all_ids[index] = set()
-        self.all_ids[index] = self.all_ids[index].union(set(ids))
-
-    def _index_name(self, index) -> str:
-        index = _sanitize_index_name(index) or self.index
-        # self.index = index  # TODO maybe not needed
+    def _index(self, index) -> str:
```
Comment (jamescalam) [inline haystack/document_stores/pinecone.py:205]: why the change from `_index_name` to `_index`? I assumed people when reading may read `index` as being the actual `Index` object rather than the index name string? Minor question, shouldn't slow down merging

File: haystack/document_stores/pinecone.py [hunk 2]
```diff
@@ -280,33 +404,73 @@ def get_document_count(
...
-                f"Index named '{index}' does not exist. Try reinitializing PineconeDocumentStore() and running "
-                f"'update_embeddings()' to create and populate an index."
-            )
+        # add `doc_type` value if specified, otherwise default is related to documents without embeddings
+        types_metadata = {type_metadata or self.document_without_embedding_metadata}
+        if not type_metadata and not only_documents_without_embedding:
+            # add `doc_type` related to documents with embeddings
+            types_metadata.add(self.document_with_embedding_metadata)
 
-        pinecone_syntax_filter = LogicalFilterClause.parse(filters).convert_to_pinecone() if filters else None
```
Comment (jamescalam) [inline haystack/document_stores/pinecone.py:297]: this is removed because we apply the filters in `_get_vector_count`?

**Review by jamescalam — APPROVED** [review]
some minor questions, but lgtm, feel free to push to haystack

File: haystack/document_stores/pinecone.py (hunk 2 above)
Comment (ivanazeljkovic) [reply haystack/document_stores/pinecone.py:297]: That's right. Common logic from `get_embedding_count` and `get_document_count` is extracted in helper method `_get_vector_count`. First step in `_get_vector_count` is to extend given filters with new ones related to `doc_type` metadata field using `types_metadata` values (if any) and after that to do parsing into Pinecone syntax.

File: haystack/document_stores/pinecone.py (hunk 1 above)
Comment (ivanazeljkovic) [reply haystack/document_stores/pinecone.py:205]: I agree with you that make more sense to be `_index_name` instead of `_index`, since it is not reference to `Index` object. 

But, this change is made just in order to be consistent in the naming with the attribute on class level, since in the constructor of `PineconeDocumentStore` class (as well as base class), `index` parameter indicates the name of the index, not `Index` reference.

**Review by jamescalam — APPROVED** [review]
hey @ivanazeljkovic — all looks good to me
